### PR TITLE
Dry run tweaks

### PR DESCRIPTION
**Pull request by annawoodard**
Collecting the suggestions on #4406:

> Do the math for the user and suggest the "correct" splitting to get to our recommended 8 hr jobs.

Done in 9ad18a2

> Much more difficult, but allow them to change the splitting in the config file and re-split the task (or I guess delete the current task and replace it transparently)

I vote we skip this for now. I don't think the extra complication to the code is worth the slight convenience (in fact, I think it might cause confusion.)

> Implement warnings in the case of memory estimate exceeding normal envelope.

Done in 9ad18a2

> Instead of parsing the dagman input file inside getDryRunOpts, I would like to take the information we need from the task database and/or from some adhoc json file we upload from the TW.

I've modified the function to take the job number as an input parameter instead as an intermediate measure.

> we upload the "dryrun sandbox" no matter what, and submit to dagman submitter only if we are not in the "dryrun" mode.

Done in dmwm/CRABServer@b5a29f3

> Allow user to skip the trial run if they only care about the splitting

I added a ‘skip-estimates’ command-line argument in 9ad18a2.

> Collect more data about how accurate the estimates are.

I added a loop to increase the number of events for psets with low time / event in  9ad18a2 based on their total runtime. Here are typical times for a simulation pset:

Events | AvgEventTime | actual runtime | dryrun estimated runtime (1280 events) | % error (1280 events) 
------------ | ------------ | ------------- | ------------ | -------------
10 | 7.3 | 124.2	 | 9392.9 | -15%
20 | 7.6 | 211.2 | 9743.2 | -19%
40 | 5.4 | 265.9	 | 6855.8 | 16%
80 | 6.5 | 572.7	 | 8365.7 | -3%
160 | 6.5 | 1110.8 | 8312.0 | -2%
320 | 6.3 | 2050.8 | 8001.9 | 2%
640 | 6.5 | 4195.2 | 8276.3 | -1%
1280 | 6.3	 | 8160.6 | 8111.5 | 1%

I've tested that the algorithm executes one 10-event file run for the pset above (~15% error). For a faster-running slimming pset (AvgEventTime ~0.008 s / event), the algorithm executes two iterations, stopping at ~6000 events with a total runtime of ~45 seconds, corresponding to ~25% error. To reduce this error, we could increase the trial time limit. There might also be some more sophisticated ways of doing this-- subtracting the file open time, for example, if someone knows which quantity in the FJR is the correct one-- but I think this is probably close enough for our purposes (and I added an explicit warning about possible issues with the estimates in 9ad18a2.)

### CODE REVIEW

File: src/python/CRABClient/ClientUtilities.py [hunk 1]
```diff
@@ -619,8 +619,14 @@ def setSubmitParserOptions(parser):
                            dest = 'dryrun',
                            default = False,
                            action = 'store_true',
-                           help = "Do not actually submit the task; instead, return how many jobs this task would create, along with processing time estimates.")
+                           help = "Do not actually submit the task; instead, return how many jobs this task would create, "\
+                                  "along with processing time and memory consumption estimates.")
 
+    parser.add_option('--skip-estimates',
```
Comment (mmascher) [inline src/python/CRABClient/ClientUtilities.py]: @annawoodard , can you document this in the twiki for the commands (I'll let @AndresTanasijczuk tell you which is the right URL :)).

File: src/python/CRABClient/Commands/submit.py [hunk 2]
```diff
@@ -333,57 +332,109 @@ def executeTestRun(self, filecacheurl):
...
+            totalJobSeconds = 0
+            maxSeconds = 25
+            while totalJobSeconds < maxSeconds:
+                opts = getCMSRunAnalysisOpts('Job.submit', 'RunJobs.dag', job=1, events=events)
+
+                s = subprocess.Popen(['sh', 'CMSRunAnalysis.sh'] + opts, env=env, stderr=subprocess.PIPE, stdout=subprocess.PIPE)
+                out, err = s.communicate()
+                self.logger.debug(out)
+                if s.returncode != 0:
+                    raise ClientException('Dry run failed to execute local test run:\n StdOut: %s\n StdErr: %s' % (err, out))
```
Comment (mmascher) [inline src/python/CRABClient/Commands/submit.py]: Invert `out` and `err` please

**Comment by mmascher** [comment]
The pull looks good to me, I added a few minor comments. I will merge this and try to run it once you update it. thanks

**Comment by AndresTanasijczuk** [comment]
Please add documentation about `--skip-estimates` in https://twiki.cern.ch/twiki/bin/view/CMSPublic/CRAB3Commands#crab_submit_dryrun

**Pull request merged**